sprockets/mixins/amqp/amqp.py

```diff
@@ -0,0 +1,210 @@
+        :param str expiration: Message expiration specification
+        :param str message_id: Application message identifier
+        :param int timestamp: Message timestamp
+        :param str message_type: Message type name
+        :param str user_id: Creating user id
+        :param str app_id: Creating application id
+        :rtype: pika.BasicProperties
+
+        """
+        return pika.BasicProperties(content_type, content_encoding, headers,
+                                    delivery_mode, priority, correlation_id,
+                                    reply_to, expiration, message_id,
+                                    timestamp, message_type, user_id, app_id)
+
+    def _publish_message(self, exchange, routing_key, message, properties):
+        """Publish the message to RabbitMQ
+
+        :param str exchange: The exchange to publish to
+        :param str routing_key: The routing key to publish with
+        :param str message: The message body
+        :param pika.BasicProperties: The message properties
+
+        """
+        LOGGER.debug('Publishing message: %s to exchange: %s with '
+                     'routing_key: %s and properties: %s',
+                     message, exchange, routing_key, properties)
+        if not self.connection or not self.channel:
+            LOGGER.error('Unable to connect to Rabbit')
+            raise pika.exceptions.AMQPError()
+        AMQPMixin.channel.basic_publish(exchange, routing_key, message,
+                                        properties)
+
+    @property
+    def _rabbit_params(self):
+        """Return a pika URLParameters object using the :envvar:`AMQP`
+        environment variable which identifies the Rabbit MQ server as a URL
+        ready for :class:`pika.connection.URLParameters`.
+
+        :rtype: pika.URLParameters
+
+        """
+        LOGGER.debug('URLParameters: %s',
+                     pika.URLParameters(os.environ['AMQP']))
+        return pika.URLParameters(os.environ['AMQP'])
+
+    def on_conn_close(self, reply_code, reply_text):
+        """Called when RabbitMQ has been connected to.
+
+        :param int reply_code: The code for the disconnect
+        :param str reply_text: The disconnect reason
+
+        """
+        LOGGER.warning('RabbitMQ has disconnected (%s): %s',
+                       reply_code, reply_text)
+        AMQPMixin.connection = None
+        self.amqp_ready.clear()
+        AMQPMixin.amqp_state = AMQP_STATES['disconnected']
+        self._connect_to_rabbitmq()
+
+    @gen.coroutine
+    def on_conn_open(self, connection):
+        """Called when RabbitMQ has been connected to.
+
+        :param pika.connection.Connection connection: The pika connection
+
+        """
+        LOGGER.info('RabbitMQ has created a connection')
+        AMQPMixin.connection = connection
+        AMQPMixin.connection.add_on_close_callback(self.on_conn_close)
+        yield self._create_new_channel()
+        self.amqp_ready.set()
```

Comment:
Because you can't add an on_close_callback to the channel until it is established